Refactor: Optimized the code and implemented countdown feature in `/instance` #6

YonLiud opened the pull request:
Read commit messages for details

YonLiud committed:
refactor: Optimize file copying for image creation

Replaced the use of the 'dd' command with shutil.copyfile for creating image copies. This change avoids the overhead of spawning external processes and improves maintainability by using Python's built-in file operations.

- Improved code readability and performance by using shutil.copyfile
- Simplified path handling with os.path.join
- Added progress messages for better tracking

YonLiud committed:
docs: Modify .gitignore to NOT upload any image files except original image & Reduce comments in __init__.py

- Add 18*.img to .gitignore
- Improve readability by reducing spaces

YonLiud committed:
feat: Misc changes to template files

- Add timer countdown to open session in `/instance` (W.I.P)
- Make button in `/` redirect to `/instance`

YonLiud committed:
docs: Add a warning to README.md

- Warning is necessary for this issue

YonLiud committed:
feat: Use cleanup_launcher when closing the app

- Now when the application closes, part of the flask's teardown wrapper
  it will remove all left-over images

YonLiud committed:
Fix: Read details

- Fixed not all files being removed on removal
- Added more print functions for logging
- Implemented signal for CTRL+C detection and auto image removal
- Removed flask termination appcontext cuz idk how it works
- Lost my sanity working on this with NeoVim... what am I doing wrong
  with my life? ...

YonLiud committed:
docs: Remove images from .gitignore

- Since files remove themselves now, no need to ignore them

s1nisteR commented on app/__init__.py:
```diff
@@ -57,26 +52,21 @@ def create_app(test_config=None):
 if __name__ == '__main__':
     app = create_app()
-    app.run(debug=True)
+    app.run(debug=True, host='0.0.0.0', port=5000)
```
We will never bind to `0.0.0.0` because the flask app will only run locally. Instead, nginx will be public-facing and will proxy all requests to the locally running flask app.

YonLiud replied:
Fixed in [65f9f11](https://github.com/omen-osdev/omen-demo/pull/6/commits/65f9f119aacbf066362251b017d8df30ef7a990b)

s1nisteR commented on app/__init__.py:
```diff
@@ -57,26 +52,21 @@ def create_app(test_config=None):
-    
     from .instance import init_launcher
-    init_launcher()    # Init the instance launcher
+    init_launcher()    # Init the instance launcher #? Should this be done here? Maybe an HTTP Method should be used to start the instance
```
Right now there are no plans to call `init_launcher()` from some sort of an HTTP API endpoint because we assume that the instance launcher is ready to go, right after we start the flask app. Maybe in future we could have some sort of an admin panel and then possibly have a HTTP method to ready the launcher.

YonLiud replied:
👍

s1nisteR commented on README.md:
````diff
@@ -18,7 +18,7 @@ pip install -r requirements.txt
-TODO
-```
+> [!WARNING]
+> Currently the app won't destroy the Docker container after the user closes the tab.</br>
````
Should be reworded. The app does indeed destroy the containers after INSTANCE_LIMIT amount of time is reached. It's just that if you close the flask app and the containers are still running(because possibly the app crashed or you did a Ctrl-c and the instance runtime did not reach the "shutdown limit" yet), then even if the flask app closes it does not shutdown the currently running containers. 

Tabs currently don't correspond to closing containers because for example:
- User launches new instance (with INSTANCE_LIMIT = 1 (minute) )
- A server-side session is created for 1 min.
- A container is created for 1 min.
- While the session exists, the container exists. User can refresh, or even close and reopen tabs. Because the cookie for the session exists in user's browser, the session runs till 1 min is reached. Similar story for the container.

@TretornESP  should we change this behavior? Should refreshing page or closing the tab be counted as shutting down the container? On one hand, it means more available containers per server. On the other hand, it means too many instance launching and shutdowns. (Imagine someone spamming refresh)

YonLiud replied:
Noted, I will keep the conversation open tho so Tretorn can have time to reply here

s1nisteR commented on README.md:
````diff
@@ -18,7 +18,7 @@ pip install -r requirements.txt
+> [!WARNING]
+> Currently the app won't destroy the Docker container after the user closes the tab.</br>
+> To manually stop the container, run:<br/>```docker ps``` to get the container ID,</br>
+> and then ```docker stop <container_id>``` to stop the container.
````
use `docker kill <container_id>` instead. 
Sometimes Omen hangs and `docker stop` does not work immediately.

YonLiud replied:
Fixed at [1f05830](https://github.com/omen-osdev/omen-demo/pull/6/commits/1f05830017e5a48058940999cda00343b29a9bb0)

s1nisteR commented on app/instance.py:
```diff
@@ -33,20 +34,77 @@ def docker_image_exists(client):
+        return
+    
+    # Create images for each port (example: 8080.img, 8081.img, etc.)
+    for port in range(8080, 8085):  # Example ports
```
Image creation should make use of `INSTANCE_PORT_NUMBER_BEGIN` and `MAX_INSTANCES` for the possible port numbers. These are defined near the top of the `instance.py` file.

Also `INSTANCE_PORT_NUMBER_BEGIN = 18200` was used because a large chunk of the ports in this range are unofficial and rarely used by most applications. Comparing it to ports like `8080, 8081, etc`, these are quite often used for web-servers and hence might cause issues in future, with the port already being occupied when starting the flask server.

Just do: 
```python
for port in available_ports:
```

`available_ports` is a list of all usable ports, already defined for you near the top of the file.

YonLiud replied:
oops, haven't notices the available_ports definition 😝

YonLiud replied:
Fixed at [ac3cfef](https://github.com/omen-osdev/omen-demo/pull/6/commits/ac3cfef7648ce02816640f9248c13c6cea95563c)

YonLiud committed:
request: remove pre-defined ip for flask app

- Removed default public IP from app ( The app will be used using a
  public nginx proxy instead )

YonLiud committed:
docs: request: "Re-word" warning message

- Rewrite warning message
- Change docker command from `stop` to `kill`

YonLiud committed:
refactor: Use pre-defined ports definiton

- Use pre-defined ports instead of range(8080, 8085)

YonLiud commented:
Should be good to go :)

YonLiud committed:
docs: Tiny improvement to readability

- Add breakline to warning

YonLiud committed:
fix: Replace p1 and p2 with a plain p tag

- There is no p1 or p2 in html... only in header tags

YonLiud committed:
feat: timeout countdown in `/instance/`

- Added a timeout countdown paragraph tag
- TODO: Add styling

YonLiud committed:
feat: Add failsafe to js in `/instance/`

- Added a failsafe message incase countdown id wont work for some reason

YonLiud committed:
fix: Countdown now works

- Linked timeout.js to instance.html
- Got countdown to work

YonLiud commented:
Live countdown will be present like so:
![image](https://github.com/user-attachments/assets/0c3d57ed-cfde-450d-9ec8-b7e66b31508a)

TODO: Add styling to this countdown

YonLiud committed:
git: Ignore all image files

- Ignore all .img files to prevent accidental uploads of instances
- Still upload original omen.img

s1nisteR committed:
Saving your SSD when debugging

s1nisteR committed:
Reduced time between limit checks

s1nisteR committed:
Added instance cleanup functions and fixed bug in port dictionary

s1nisteR committed:
Shutdown does not get stuck anymore at exit(0)

s1nisteR committed:
Updated README.md

s1nisteR commented:
Nice work! Everything looks perfect now. Merging and closing PR.
